// vue-ssr-webpack-plugin/lib/client.js
var hash = require('hash-sum')
var ref = require('./util');
var isJS = ref.isJS;

// const debug = (file, obj) => {
//   require('fs').writeFileSync(__dirname + '/' + file, JSON.stringify(obj, null, 2))
// }

module.exports = (function () {
  function VueSSRClientPlugin (options) {
  if ( options === void 0 ) options = {};

    this.options = Object.assign({
      filename: 'vue-ssr-client-manifest.json',
    }, options)
  }

  VueSSRClientPlugin.prototype.apply = function apply (compiler) {
    var this$1 = this;

    compiler.plugin('emit', function (compilation, cb) {
      var stats = compilation.getStats().toJson()

      var allFiles = stats.assets
        .map(function (a) { return a.name; })

      var initialScripts = Object.keys(stats.entrypoints)
        .map(function (name) { return stats.entrypoints[name].assets; })
        .reduce(function (assets, all) { return all.concat(assets); }, [])
        .filter(isJS)

      var asyncScripts = allFiles
        .filter(isJS)
        .filter(function (file) { return initialScripts.indexOf(file) < 0; })

      var manifest = {
        publicPath: stats.publicPath,
        all: allFiles,
        initial: initialScripts,
        async: asyncScripts,
        modules: { /* [identifier: string]: Array<index: number> */ }
      }

      var assetModules = stats.modules.filter(function (m) { return m.assets.length; })
      var fileToIndex = function (file) { return manifest.all.indexOf(file); }
      stats.modules.forEach(function (m) {
        // ignore modules duplicated in multiple chunks
        if (m.chunks.length === 1) {
          var cid = m.chunks[0]
          var chunk = stats.chunks.find(function (c) { return c.id === cid; })
          var files = manifest.modules[hash(m.identifier)] = chunk.files.map(fileToIndex)
          // find all asset modules associated with the same chunk
          assetModules.forEach(function (m) {
            if (m.chunks.some(function (id) { return id === cid; })) {
              files.push.apply(files, m.assets.map(fileToIndex))
            }
          })
        }
      })

      // debug('stats.json', stats)
      // debug('client-manifest.json', manifest)

      var json = JSON.stringify(manifest, null, 2)
      compilation.assets[this$1.options.filename] = {
        source: function () { return json; },
        size: function () { return json.length; }
      }
      cb()
    })
  };

  return VueSSRClientPlugin;
}())
